File: src/main/java/net/hellomouse/progression_change/ProgressionMod.java
package net.hellomouse.progression_change;

import com.mojang.logging.LogUtils;
import net.hellomouse.progression_change.registries.*;
import net.minecraft.item.ItemGroup;
import net.minecraft.item.ItemGroups;
import net.minecraft.item.ItemStack;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.text.Text;
import net.minecraft.util.Identifier;
import net.minecraftforge.common.TierSortingRegistry;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import net.minecraftforge.registries.DeferredRegister;
import net.minecraftforge.registries.RegistryObject;
import org.slf4j.Logger;

import java.util.List;

@Mod(ProgressionMod.MODID)
public class ProgressionMod {
    public static final String MODID = "progression_change";
    public static final Logger LOGGER = LogUtils.getLogger();
    public static final DeferredRegister<ItemGroup> CREATIVE_TAB_REG = DeferredRegister.create(RegistryKeys.ITEM_GROUP, MODID);
    public static final RegistryObject<ItemGroup> CREATIVE_TAB = CREATIVE_TAB_REG.register(MODID, () -> ItemGroup.builder()
            .displayName(Text.translatable("itemGroup." + MODID + ".creative_tab"))
            .icon(() -> new ItemStack(ProgressionModItemRegistry.COPPER_PICKAXE.get()))
            .withTabsBefore(ItemGroups.SPAWN_EGGS)
            .entries((enabledFeatures, output) -> {
                ProgressionModItemRegistry.addItemToCreativeTab(output);
            })
            .build());

    public ProgressionMod(FMLJavaModLoadingContext context) {
        LOGGER.info("Loading Progression Mod Config...");

        IEventBus modEventBus = context.getModEventBus();
        ProgressionModBlockRegistry.DEF_REG.register(modEventBus);
        ProgressionModItemRegistry.VANILLA_ITEMS.register(modEventBus);
        ProgressionModItemRegistry.DEF_REG.register(modEventBus);
        ProgressionModEntityRegistry.DEF_REG.register(modEventBus);
        ProgressionModScreenHandlerRegistry.DEF_REG.register(modEventBus);
        ProgressionModBlockEntityRegistry.DEF_REG.register(modEventBus);
        ProgressionModLootTypeRegistry.FUNC_DEF_REG.register(modEventBus);
        ProgressionModLootTypeRegistry.COND_DEF_REG.register(modEventBus);
        ProgressionModRecipeRegistry.DEF_REG.register(modEventBus);
        ProgressionModRecipeRegistry.TYPE_DEF_REG.register(modEventBus);
        CREATIVE_TAB_REG.register(modEventBus);
        TierSortingRegistry.registerTier(ProgressionModToolMaterials.COPPER,
                ProgressionModToolMaterials.COPPER_ID,
                List.of(new Identifier("stone")),
                List.of(new Identifier("iron"))
        );
        TierSortingRegistry.registerTier(ProgressionModToolMaterials.FLINT,
                ProgressionModToolMaterials.FLINT_ID,
                List.of(),
                List.of(new Identifier("wood"))
        );
        TierSortingRegistry.registerTier(ProgressionModToolMaterials.BONE,
                ProgressionModToolMaterials.BONE_ID,
                List.of(),
                List.of(new Identifier("wood"))
        );
    }


    public static Identifier of(String path) {
        return Identifier.of(MODID, path);
    }
}

File: src/main/kotlin/net/hellomouse/xeno_early_start/ProgressionMod.kt
package net.hellomouse.xeno_early_start

import com.mojang.logging.LogUtils
import me.shedaniel.cloth.clothconfig.shadowed.com.moandjiezana.toml.Toml
import net.hellomouse.xeno_early_start.registries.*
import net.minecraft.client.MinecraftClient
import net.minecraft.client.gui.screen.Screen
import net.minecraft.item.ItemGroup
import net.minecraft.item.ItemGroups
import net.minecraft.item.ItemStack
import net.minecraft.registry.RegistryKeys
import net.minecraft.text.Text
import net.minecraft.util.Identifier
import net.minecraftforge.client.ConfigScreenHandler.ConfigScreenFactory
import net.minecraftforge.common.TierSortingRegistry
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.loading.FMLPaths
import net.minecraftforge.registries.DeferredRegister
import net.minecraftforge.registries.RegistryObject
import org.slf4j.Logger
import thedarkcolour.kotlinforforge.forge.LOADING_CONTEXT
import thedarkcolour.kotlinforforge.forge.MOD_BUS
import java.io.FileNotFoundException
import java.util.function.Supplier

@Mod(ProgressionMod.MODID)
class ProgressionMod {
    init {
        LOGGER.info("Loading Progression Mod Config...")

        ProgressionModBlockRegistry.DEF_REG.register(MOD_BUS)
        ProgressionModItemRegistry.VANILLA_ITEMS.register(MOD_BUS)
        ProgressionModItemRegistry.DEF_REG.register(MOD_BUS)
        ProgressionModEntityRegistry.DEF_REG.register(MOD_BUS)
        ProgressionModScreenHandlerRegistry.DEF_REG.register(MOD_BUS)
        ProgressionModBlockEntityRegistry.DEF_REG.register(MOD_BUS)
        ProgressionModLootTypeRegistry.FUNC_DEF_REG.register(MOD_BUS)
        ProgressionModLootTypeRegistry.COND_DEF_REG.register(MOD_BUS)
        ProgressionModRecipeRegistry.DEF_REG.register(MOD_BUS)
        ProgressionModRecipeRegistry.TYPE_DEF_REG.register(MOD_BUS)
        XenoProgressionModParticleRegistry.DEF_REG.register(MOD_BUS)
        XenoEarlyStartStatusEffectRegistry.DEF_REG.register(MOD_BUS)
        CREATIVE_TAB_REG.register(MOD_BUS)
        TierSortingRegistry.registerTier(
            ProgressionModToolMaterials.COPPER,
            ProgressionModToolMaterials.COPPER_ID,
            listOf<Any?>(ofMinecraft("stone")),
            listOf<Any?>(ofMinecraft("iron"))
        )
        TierSortingRegistry.registerTier(
            ProgressionModToolMaterials.FLINT,
            ProgressionModToolMaterials.FLINT_ID,
            mutableListOf<Any?>(),
            listOf<Any?>(ofMinecraft("wood"))
        )
        TierSortingRegistry.registerTier(
            ProgressionModToolMaterials.BONE,
            ProgressionModToolMaterials.BONE_ID,
            mutableListOf<Any?>(),
            listOf<Any?>(ofMinecraft("wood"))
        )
        try {
            val file = FMLPaths.CONFIGDIR.get().resolve("xeno-early-start.toml").toFile()
            ProgressionModConfig.config =
                Toml().read(file.reader())
                    .to(ProgressionModConfig.config.javaClass)
        } catch (_: FileNotFoundException) {
            val tomlWriter = me.shedaniel.cloth.clothconfig.shadowed.com.moandjiezana.toml.TomlWriter()
            tomlWriter.write(
                ProgressionModConfig.config,
                FMLPaths.CONFIGDIR.get().resolve("xeno-early-start.toml").toFile()
            )
        }
        LOADING_CONTEXT.registerExtensionPoint(
            ConfigScreenFactory::class.java,
            Supplier {
                ConfigScreenFactory { mc: MinecraftClient?, prevScreen: Screen? ->
                    XenoEarlyStartConfigGui.configBuilder.build()
                }
            }
        )


    }


    companion object {
        const val MODID: String = "xeno_early_start"
        val LOGGER: Logger = LogUtils.getLogger()
        val CREATIVE_TAB_REG: DeferredRegister<ItemGroup> =
            DeferredRegister.create<ItemGroup>(RegistryKeys.ITEM_GROUP, MODID)

        @Suppress("UNUSED")
        val CREATIVE_TAB: RegistryObject<ItemGroup> = CREATIVE_TAB_REG.register<ItemGroup>(MODID, Supplier {
            ItemGroup.builder()
                .displayName(Text.translatable("itemGroup.$MODID.creative_tab"))
                .icon { ItemStack(ProgressionModItemRegistry.COPPER_PICKAXE.get()) }
                .withTabsBefore(ItemGroups.SPAWN_EGGS)
                .entries { enabledFeatures: ItemGroup.DisplayContext?, output: ItemGroup.Entries ->
                    ProgressionModItemRegistry.addItemToCreativeTab(output)
                }
                .build()
        })

        fun of(path: String): Identifier {
            return Identifier.of(MODID, path) ?: throw IllegalArgumentException("$MODID:$path is not valid")
        }

        fun ofMinecraft(path: String): Identifier {
            return Identifier.of("minecraft", path) ?: throw IllegalArgumentException("$MODID:$path is not valid")
        }
    }
}
